File: src/maths_parser_detroix23/structures/tokens.py
"""
# Python mathematics parser.
src/maths_parser_detroix23/structures/tokens.py
"""

from typing import Final

from maths_parser_detroix23.structures import defaults

tokens: list[Token] = []
""" Register all `Token`s: one centralized access. """

class Token:
	"""
	# `Token` trait for all symbols.
	Holds:
	- `representations`: `list[str]`
	"""
	representations: list[str]

	def __init__(self, representations: list[str]) -> None:
		"""
		Initialize a basic `Token` with its `representations` (at least one).

		Also register it`self` into `tokens`.
		"""
		self.representations = representations

		tokens.append(self)
	
	def __str__(self) -> str:
		"""
		Returns the first `representations` of the `Token` 
		"""
		return self.representations[0]

	def __repr__(self) -> str:
		"""
		Returns a string: `Token(representations)`.
		"""
		return f"Token(representations={self.representations})"


UNKNOWN: Final[Token] = Token([defaults.UNKNOWN_TOKEN])
BLOCK_OPENING: Final[Token] = Token(["("])
BLOCK_CLOSING: Final[Token] = Token([")"])
	

File: rust/src/basic/tokens.rs
//! # Python mathematics parser.
//! /src/defaults/tokens.rs

use crate::modules::vec_string;
use crate::structures::tokens;

pub fn generate_operations() -> Vec<tokens::Token> {
	vec![
		tokens::Token::new(
			"Addition".to_string(),
			vec_string(&["+"]),
			tokens::TokenType::FUNCTION,
		),
		tokens::Token::new(
			"Subtraction".to_string(),
			vec_string(&["-"]),
			tokens::TokenType::FUNCTION,
		),
		tokens::Token::new(
			"Multiplication".to_string(),
			vec_string(&["*"]),
			tokens::TokenType::FUNCTION,
		),
		tokens::Token::new(
			"Division".to_string(),
			vec!["/".to_string()],
			tokens::TokenType::FUNCTION,
		),
		tokens::Token::new(
			"Exponentiation".to_string(),
			vec_string(&["^", "**"]),
			tokens::TokenType::FUNCTION,
		),
	]
}